nixos/modules/installer/netboot/kexec-boot.sh

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env bash
+set -e
+
+kexecTree=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)
+initrd=$kexecTree/initrd
+kernelParams=$(<"$kexecTree/kernel-params")
+
+usage() {
+    printf >&2 'usage: %s [%s] [%s]\n' "$0" \
+        '--ssh-authorized-keys <file>' \
+        '--append <cmdline>'
+    exit 2
+}
+
+while (( $# )); do
+    arg=$1; shift 1
+    case "$arg" in
+        --ssh-authorized-keys)
+            (( $# )) || usage
+            authorizedKeysFile=$1; shift 1
+            ;;
+        --append)
+            (( $# )) || usage
+            kernelParams="$kernelParams $1"; shift 1
+            ;;
+        *)
+            usage
+            ;;
+    esac
+done
+
+if ! kexec --version &>/dev/null; then
+    printf >&2 'kexec not found: please install kexec-tools\n'
+    exit 1
+fi
+
+if (( EUID != 0 )); then
+    printf >&2 'kexec-boot must be run as root\n'
+    exit 1
+fi
+
+cleanup() {
+    if [[ -v workDir ]]; then
+        rm -rf "$workDir"
+    fi
+}
+
+trap cleanup EXIT
+
+if [[ -v authorizedKeysFile ]]; then
+    workDir=$(mktemp -d --tmpdir nixos-kexec-initrd.XXXXXXXXXX)
+    keysDir=$workDir/overlay/etc/ssh/authorized_keys.d
+    mkdir -p -- "$keysDir"
+    install --mode=0644 -- "$authorizedKeysFile" "$keysDir/nixos"
+    {
+        (
+            cd -- "$workDir"
+            find overlay -print0 | cpio -o -H newc --null --quiet
+        )
+        cat -- "$initrd"
+    } >"$workDir/initrd"
```

Comment:
I recently switched to just append to the original $initrd here. This was to work-around an out-of-memory event. Also its a bit faster: https://github.com/nix-community/nixos-images/blob/0092c8af6cd2ae4079c6f2f9dfa3fdaa1622c709/nix/kexec-installer/module.nix#L53

Reply:
Strangely enough, prepending is actually required here as-is; initially I was appending the overlay cpio, but I found that the kernel wouldn't include it (at least on my AArch64 test system) when appended to a zstd-compressed initrd. Compressing the overlay makes appending work, but between the choice of an additional dependency just to compress a few bytes and prepending I picked the latter. So we would have to depend on or bundle `zstd(1)` to do this.

Between that and the fact that repeated runs (e.g. if you pick the wrong keys the first time) would conflict with each other if modifying in-place (I'm not sure how such duplicate paths are resolved by the kernel) I'm inclined to keep it as-is, but I guess it means a more annoying bootstrap on systems that can't spare ~800 MiB of either disk or RAM. (In most cases you could just delete the tarball to free up the necessary space, under the assumption that you already needed twice the storage to unpack it in the first place.)

Reply:
Yeah in some circumstances you want to kexec from netboot recovery systems, so disk space might be limited.

Reply:
Right, makes sense. How would you be extracting the tarball in the first place if you don't have approximately 2 × 800 MiB to spare initially (and could therefore delete the tarball to save the space for the copy of the initrd)? I guess you could just pipe curl into tar?